### PR TITLE
Fix Conditional Group Dependencies not working as expected

### PR DESCRIPTION
**Pull request by matthid**
https://github.com/Antaris/RazorEngine/pull/465

### CODE REVIEW

File: src/Paket.Core/Installation/InstallProcess.fs [hunk 1]
```diff
@@ -390,73 +390,79 @@ let InstallIntoProjects(options : InstallerOptions, forceTouch, dependenciesFile
...
+                | true,(v',hasCondition') ->
+                    if v' = v then
+                        traceWarnfn "Package %O is referenced through multiple groups in %s (inspect lockfile for details). To resolve this warning use a single group for this project to get a unified dependency resolution or use conditions on the groups if you know what you are doing." packageName project.FileName
+                        false
+                    else
+                        errors.Add <| sprintf "Package %O is referenced in different versions in %s (%O vs %O), (inspect the lockfile for details) to resolve this either add all dependencies to a single group (to get a unified resolution) or use a condition on both groups and control compilation yourself." packageName project.FileName v' v
+                        false
                 | _ ->
-                    dict.Add(packageName,v)
+                    dict.Add(packageName,(v,hasCondition))
```
Comment (matthid) [inline src/Paket.Core/Installation/InstallProcess.fs]: here are the relevant changes...

File: src/Paket.Core/Installation/InstallProcess.fs [hunk 2]
```diff
@@ -390,73 +390,79 @@ let InstallIntoProjects(options : InstallerOptions, forceTouch, dependenciesFile
...
-                | true,v' -> 
-                    if v' = v then false else
-                    errors.Add <| sprintf "Package %O is referenced in different versions in %s (%O vs %O), check the lock-file." packageName project.FileName v' v
-                    false
+                | true,(v',true) when hasCondition ->
+                    true
+                | true,(v',hasCondition') ->
+                    if v' = v then
+                        traceWarnfn "Package %O is referenced through multiple groups in %s (inspect lockfile for details). To resolve this warning use a single group for this project to get a unified dependency resolution or use conditions on the groups if you know what you are doing." packageName project.FileName
+                        false
```
Comment (matthid) [inline src/Paket.Core/Installation/InstallProcess.fs]: We can argue about this warning, but I think it's the correct thing to do...

Comment (matthid) [inline src/Paket.Core/Installation/InstallProcess.fs]: /cc @agross This basically means you get this warning on your example from https://github.com/fsprojects/Paket/issues/1458. IMHO we should not encourage this kind of paket group usage.

**Comment by matthid** [comment]
Apparently we broke the CI with https://github.com/fsprojects/Paket/commit/5e9c716b9d3a0809504f1fd54394c024731e0d53 (because the error message was changed only in the code, not in the test). Maybe we should have a similar test running on travis as well?
For now I'll just go ahaid and fix it as part of this PR...

File: src/Paket.Core/Dependencies/PackageResolver.fs [hunk 3]
```diff
@@ -170,27 +170,24 @@ module Resolution =
...
                 | [] -> errorReport
-                | hd::tl ->
-                    let vr = formatVR hd.VersionRequirement
-                    let pr = formatPR hasPrereleases hd.VersionRequirement
-
-                    match hd.Parent with
-                    | DependenciesFile _ ->
-                        loop tl (errorReport.AppendLinef "   - Dependencies file requested package %O: %s%s" req.Name vr pr)
-                    | Package (parentName,version,_) ->
-                        loop tl (errorReport.AppendLinef  "   - %O %O requested package %O: %s%s" parentName version req.Name vr pr)
```
Comment (matthid) [inline src/Paket.Core/Dependencies/PackageResolver.fs]: `req` here is a bug, right?

**Comment by forki** [comment]
Probably

Am 14.05.2017 11:09 vorm. schrieb "Matthias Dittrich" <
notifications@github.com>:

File: src/Paket.Core/Installation/InstallProcess.fs (hunk 2 above)
Comment (matthid) [inline src/Paket.Core/Installation/InstallProcess.fs]: See https://github.com/Antaris/RazorEngine/pull/465#issuecomment-301315936

**Comment by matthid** [comment]
AppVeyor seems kind of slow today, but I don't think the timeouts are related to the changes in this PR...

**Comment by matthid** [comment]
Or we introduced some change which makes everything slower recently ...

**Pull request merged**